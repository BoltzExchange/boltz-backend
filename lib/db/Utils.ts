import { DatabaseType } from './Database';

type Queries = {
  [DatabaseType.SQLite]: string;
  [DatabaseType.PostgreSQL]: string;
};

export { Queries };
