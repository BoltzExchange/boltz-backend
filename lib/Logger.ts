import { blue, cyan, green, magenta, red, yellow } from 'colors/safe';
import winston from 'winston';
import LokiTransport from 'winston-loki';
import { name } from '../package.json';
import { getTsString } from './Utils';

class Logger {
  public static readonly disabledLogger = new Logger(
    '',
    undefined,
    undefined,
    undefined,
    true,
  );

  private readonly loki?: LokiTransport;

  constructor(
    level: string,
    filename?: string,
    lokiHost?: string,
    lokiNetwork?: string,
    private disabled = false,
  ) {
    if (disabled) {
      return;
    }

    const transports: winston.transport[] = [
      new winston.transports.Console({
        format: this.getLogFormat(true),
      }),
    ];

    if (filename) {
      transports.push(
        new winston.transports.File({
          filename,
          format: this.getLogFormat(false),
        }),
      );
    }

    const lokiEnabled = [lokiHost, lokiNetwork].every(
      (val) => val !== undefined,
    );

    if (lokiEnabled) {
      this.loki = new LokiTransport({
        host: lokiHost!,
        labels: {
          job: name,
          network: lokiNetwork,
        },
      });
      transports.push(this.loki!);
    }

    winston.configure({
      level,
      transports,
    });

    if (!lokiEnabled) {
      this.warn('Disabled loki logger because of invalid configuration');
    } else {
      this.debug('Enabled loki logger');
    }
  }

  public close = async () => {
    if (this.loki !== undefined) {
      await this.loki.flush();
      await (this.loki as any).close();
    }
  };

  private getLogFormat = (colorize: boolean) => {
    return winston.format.printf(
      (info) =>
        `${getTsString()} ${this.getLevel(info.level, colorize)}: ${
          info.message
        }`,
    );
  };

  private getLevel = (level: string, colorize: boolean) => {
    if (colorize) {
      switch (level) {
        case 'error':
          return red(level);
        case 'warn':
          return yellow(level);
        case 'info':
          return green(level);
        case 'verbose':
          return cyan(level);
        case 'debug':
          return blue(level);
        case 'silly':
          return magenta(level);
      }
    }
    return level;
  };

  public error = (message: string): void => {
    this.log('error', message);
  };

  public warn = (message: string): void => {
    this.log('warn', message);
  };

  public info = (message: string): void => {
    this.log('info', message);
  };

  public verbose = (message: string): void => {
    this.log('verbose', message);
  };

  public debug = (message: string): void => {
    this.log('debug', message);
  };

  public silly = (message: string): void => {
    this.log('silly', message);
  };

  private log = (level: string, message: string) => {
    if (!this.disabled) {
      winston.log(level, message);
    }
  };
}

export default Logger;
