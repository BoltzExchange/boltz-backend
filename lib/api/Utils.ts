import { Request, Response } from 'express';
import Logger from '../Logger';
import { getHexBuffer } from '../Utils';
import MarkedSwapRepository from '../db/repositories/MarkedSwapRepository';
import CountryCodes from '../service/CountryCodes';
import ServiceErrors from '../service/Errors';
import Errors from './Errors';

type ApiArgument = {
  name: string;
  type: 'string' | 'number' | 'boolean' | 'object';
  hex?: boolean;
  optional?: boolean;
};

// Some endpoints are getting spammed on mainnet, so we don't log the warnings for them
const errorsNotToLog: any[] = [
  ServiceErrors.SWAP_NO_LOCKUP().message,
  ServiceErrors.ETHEREUM_NOT_ENABLED().message,
];

/**
 * Validates that all required arguments were provided in the body correctly
 *
 * @returns the validated arguments
 */
export const validateRequest = (
  body: Record<string, any>,
  argsToCheck: ApiArgument[],
) => {
  const response: any = {};

  argsToCheck.forEach((arg) => {
    const value = body[arg.name];

    if (value !== undefined) {
      if (typeof value === arg.type) {
        if (arg.hex && value !== '') {
          const buffer = getHexBuffer(value);

          if (buffer.length === 0) {
            throw Errors.COULD_NOT_PARSE_HEX(arg.name);
          }

          response[arg.name] = buffer;
        } else {
          response[arg.name] = value;
        }
      } else {
        throw Errors.INVALID_PARAMETER(arg.name);
      }
    } else if (!arg.optional) {
      throw Errors.UNDEFINED_PARAMETER(arg.name);
    }
  });

  return response;
};

export const errorResponse = (
  logger: Logger,
  req: Request,
  res: Response,
  error: unknown,
  statusCode = 400,
): void => {
  if (typeof error === 'string') {
    writeErrorResponse(logger, req, res, statusCode, { error });
  } else {
    const errorObject = error as any;

    // Bitcoin Core related errors
    if (errorObject.details) {
      writeErrorResponse(logger, req, res, statusCode, {
        error: errorObject.details,
      });
      // Custom error when broadcasting a refund transaction fails because
      // the locktime requirement has not been met yet
    } else if (errorObject.timeoutBlockHeight) {
      writeErrorResponse(logger, req, res, statusCode, error);
      // Everything else
    } else {
      writeErrorResponse(logger, req, res, statusCode, {
        error: errorObject.message,
      });
    }
  }
};

export const successResponse = (res: Response, data: unknown) => {
  setContentTypeJson(res);
  res.status(200);

  if (typeof data === 'object') {
    res.json(data);
  } else {
    res.write(data);
    res.end();
  }
};

export const createdResponse = (res: Response, data: unknown) => {
  setContentTypeJson(res);
  res.status(201).json(data);
};

export const writeErrorResponse = (
  logger: Logger,
  req: Request,
  res: Response,
  statusCode: number,
  error: any,
) => {
  if (!errorsNotToLog.includes(error?.error || error)) {
    logger.warn(
      `Request ${req.method} ${req.originalUrl} ${
        req.body && Object.keys(req.body).length > 0
          ? `${JSON.stringify(req.body)} `
          : ''
      }failed: ${JSON.stringify(error)}`,
    );
  }

  setContentTypeJson(res);
  res.status(statusCode).json(error);
};

export const setContentTypeJson = (res: Response) => {
  res.set('Content-Type', 'application/json');
};

export const checkPreimageHashLength = (preimageHash: Buffer) => {
  if (preimageHash.length !== 32) {
    throw `invalid preimage hash length: ${preimageHash.length}`;
  }
};

export const markSwap = async (
  countryCodes: CountryCodes,
  ip: string,
  swapId: string,
) => {
  if (!countryCodes.isRelevantCountry(countryCodes.getCountryCode(ip))) {
    return;
  }

  await MarkedSwapRepository.addMarkedSwap(swapId);
};
