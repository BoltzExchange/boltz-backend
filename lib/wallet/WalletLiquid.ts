import { address, networks, payments } from 'liquidjs-lib';
import { Payment } from 'liquidjs-lib/src/payments';
import { Slip77Interface } from 'slip77';
import Logger from '../Logger';
import { CurrencyType } from '../consts/Enums';
import Wallet from './Wallet';
import WalletProviderInterface from './providers/WalletProviderInterface';

class WalletLiquid extends Wallet {
  constructor(
    logger: Logger,
    walletProvider: WalletProviderInterface,
    private readonly slip77: Slip77Interface,
  ) {
    super(logger, CurrencyType.Liquid, walletProvider);
  }

  public deriveBlindingKeyFromScript = (outputScript: Buffer) => {
    return this.slip77.derive(outputScript);
  };

  public override encodeAddress = (
    outputScript: Buffer,
    shouldBlind = true,
  ): string => {
    try {
      // Fee output of Liquid
      if (outputScript.length == 0) {
        return '';
      }

      const res = this.getPaymentFunc(outputScript)({
        output: outputScript,
        network: this.network as networks.Network,
        blindkey: shouldBlind
          ? this.deriveBlindingKeyFromScript(outputScript).publicKey!
          : undefined,
      });

      return shouldBlind ? res.confidentialAddress! : res.address!;
    } catch (error) {
      // Ignore invalid addresses
      return '';
    }
  };

  private getPaymentFunc = (outputScript: Buffer) => {
    switch (address.getScriptType(outputScript)) {
      case address.ScriptType.P2Pkh:
        return payments.p2pkh;
      case address.ScriptType.P2Sh:
        return payments.p2sh;
      case address.ScriptType.P2Wpkh:
        return payments.p2wpkh;
      case address.ScriptType.P2Wsh:
        return payments.p2wsh;
      default:
        return WalletLiquid.blindP2tr;
    }
  };

  private static blindP2tr = (
    payment: Payment,
  ): {
    address: string;
    confidentialAddress: string | undefined;
  } => {
    const addr = address.fromOutputScript(payment.output!, payment.network);
    const dec = address.fromBech32(addr);

    let confidentialAddress: string | undefined;

    if (payment.blindkey) {
      confidentialAddress = address.toBlech32(
        Buffer.concat([Buffer.from([dec.version, dec.data.length]), dec.data]),
        payment.blindkey,
        payment.network!.blech32,
        dec.version,
      );
    }

    return {
      confidentialAddress,
      address: addr,
    };
  };
}

export default WalletLiquid;
