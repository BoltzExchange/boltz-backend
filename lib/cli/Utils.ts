import { Arguments } from 'yargs';
import { address, ECPair, Transaction } from 'bitcoinjs-lib';
import { Networks, detectSwap, constructClaimTransaction, constructRefundTransaction } from 'boltz-core';
import inquire from './inquire';
import { getHexBuffer } from '../Utils';
import { OutputType, OrderSide } from '../proto/boltzrpc_pb';

export const getOrderSide = (side: string) => {
  switch (side.toLowerCase()) {
    case 'buy': return OrderSide.BUY;
    case 'sell': return OrderSide.SELL;

    default: throw `could not find order side: ${side}`;
  }
};

export const getOutputType = (type: string) => {
  switch (type.toLowerCase()) {
    case 'bech32': return OutputType.BECH32;
    case 'compatibility': return OutputType.COMPATIBILITY;

    default: return OutputType.LEGACY;
  }
};

export const getNetwork = (networkKey: string) => {
  const network = Networks[networkKey];

  if (!network) {
    throw `Could not find network: ${networkKey}`;
  }

  return network;
};

const parseParameters = (argv: Arguments) => {
  const network = getNetwork(argv.network);

  return {
    network,
    lockupTransaction: Transaction.fromHex(argv.lockup_transaction),
    redeemScript: getHexBuffer(argv.redeem_script),
    destinationScript: address.toOutputScript(argv.destination_address, network),
  };
};

const parseSwapOutput = (redeemScript: Buffer, lockupTransaction: Transaction) => {
  const swapOutput = detectSwap(redeemScript, lockupTransaction);

  if (!swapOutput) {
    throw 'Could not find swap in transaction';
  }

  return swapOutput;
};

export const parseCommands = async (inquiries: any[], argv: Arguments): Promise<Arguments> => {
  const argvLength = Object.keys(argv).length;
  if (argvLength === inquiries.length) {
    const answers = await inquire(inquiries);
    return { ...answers, ...argv };
  } else {
    return argv;
  }
};

export const claimSwap = (argv: Arguments) => {
  const { network, lockupTransaction, redeemScript, destinationScript } = parseParameters(argv);
  const claimKeys = ECPair.fromPrivateKey(getHexBuffer(argv.claim_private_key), { network });

  const swapOutput = parseSwapOutput(redeemScript, lockupTransaction);

  const claimTransaction = constructClaimTransaction(
    {
      redeemScript,
      preimage: getHexBuffer(argv.preimage),
      keys: claimKeys,
    },
    {
      txHash: lockupTransaction.getHash(),
      ...swapOutput,
    },
    destinationScript,
    argv.fee_per_byte,
  );

  return claimTransaction.toHex();
};

export const refundSwap = (argv: Arguments) => {
  const { network, lockupTransaction, redeemScript, destinationScript } = parseParameters(argv);
  const refundKeys = ECPair.fromPrivateKey(getHexBuffer(argv.refund_private_key), { network });

  const swapOutput = parseSwapOutput(redeemScript, lockupTransaction);

  const refundTransaction = constructRefundTransaction(
    refundKeys,
    redeemScript,
    argv.timeout_block_height,
    {
      txHash: lockupTransaction.getHash(),
      ...swapOutput,
    },
    destinationScript,
    argv.fee_per_byte,
  );

  return refundTransaction.toHex();
};
