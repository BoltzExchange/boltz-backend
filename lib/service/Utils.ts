import { getUnixTime } from '../Utils';
import type { Currency } from '../wallet/WalletManager';
import Errors from './Errors';
import TimeoutDeltaProvider from './TimeoutDeltaProvider';

export const getCurrency = (
  currencies: Map<string, Currency>,
  symbol: string,
) => {
  const currency = currencies.get(symbol);
  if (!currency) {
    throw Errors.CURRENCY_NOT_FOUND(symbol);
  }

  return currency;
};

export const calculateTimeoutDate = (chain: string, blocksMissing: number) => {
  const blockTime = TimeoutDeltaProvider.blockTimes.get(chain);
  if (blockTime === undefined) {
    throw Errors.CURRENCY_NOT_FOUND(chain);
  }

  return getUnixTime() + blocksMissing * blockTime * 60;
};
