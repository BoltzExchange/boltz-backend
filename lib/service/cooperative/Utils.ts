import { crypto } from 'bitcoinjs-lib';
import type { Types } from 'boltz-core';
import {
  createMusig,
  hashForWitnessV1,
  parseTransaction,
  tweakMusig,
} from '../../Core';
import { getHexString } from '../../Utils';
import type Wallet from '../../wallet/Wallet';
import type { Currency } from '../../wallet/WalletManager';
import Errors from '../Errors';
import type { PartialSignature } from './MusigSigner';

export const isPreimageValid = (
  swap: { preimageHash: string },
  preimage: Buffer,
) =>
  preimage.length === 32 &&
  getHexString(crypto.sha256(preimage)) === swap.preimageHash;

export const createPartialSignature = async (
  currency: Currency,
  wallet: Wallet,
  swapTree: Types.SwapTree,
  keyIndex: number,
  theirPublicKey: Buffer,
  theirNonce: Buffer,
  rawTransaction: Buffer | string,
  vin: number,
): Promise<PartialSignature> => {
  const tx = parseTransaction(currency.type, rawTransaction);
  if (vin < 0 || tx.ins.length <= vin) {
    throw Errors.INVALID_VIN();
  }

  const ourKeys = wallet.getKeysByIndex(keyIndex);

  const musig = createMusig(ourKeys, theirPublicKey);
  tweakMusig(currency.type, musig, swapTree);

  musig.aggregateNonces([[theirPublicKey, theirNonce]]);

  const hash = await hashForWitnessV1(currency, tx, vin);
  musig.initializeSession(hash);

  return {
    signature: Buffer.from(musig.signPartial()),
    pubNonce: Buffer.from(musig.getPublicNonce()),
  };
};
