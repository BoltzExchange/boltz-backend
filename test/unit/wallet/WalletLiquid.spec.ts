import ops from '@boltz/bitcoin-ops';
import { crypto } from 'bitcoinjs-lib';
import { toXOnly } from 'bitcoinjs-lib/src/psbt/bip371';
import { Scripts } from 'boltz-core';
import Networks from 'boltz-core/dist/lib/liquid/consts/Networks';
import { address } from 'liquidjs-lib';
import { SLIP77Factory } from 'slip77';
import * as ecc from 'tiny-secp256k1';
import Logger from '../../../lib/Logger';
import { getHexBuffer } from '../../../lib/Utils';
import WalletLiquid from '../../../lib/wallet/WalletLiquid';

describe('WalletLiquid', () => {
  const slip77 = SLIP77Factory(ecc).fromSeed(
    'test test test test test test test test test test test junk',
  );
  const wallet = new WalletLiquid(Logger.disabledLogger, {} as any, slip77);
  wallet.initKeyProvider(Networks.liquidRegtest, '', 0, {} as any);

  const publicKey = getHexBuffer(
    '03f0081c29011d63e741e4bfe2465a9e1bb203852d239f541d92dc8d9e40bdb3e6',
  );

  test.each`
    addr
    ${'AzpvTi6t8GTVxhg6tZ4vxNCdQKS9YYMMgJmxQAFYfhVvaogb6iVgTixvtv246tSbeM3zdgG1Z2ToreMt'}
    ${'CTEvqk9mbKSWnkPhF7DwqHf5X5Jx1Q25LXh4sprdqj4KRgMZTZaiGrhCCDfWrDyVqBbxUrhyCtLwgB7J'}
    ${'el1qqwxdjljvzdukcfxeammq6jktrvcurh329k7j208pm4rwe7anu2luhh7255kde06j2et2d0g5ym6yy949rx6rkp04xeav62vjp'}
  `('should derive blinding keys from script for address $addr', ({ addr }) => {
    const blindingKey = wallet.deriveBlindingKeyFromScript(
      address.toOutputScript(addr, Networks.liquidRegtest),
    );
    expect(blindingKey.privateKey).toMatchSnapshot();
  });

  test.each`
    name        | shouldBlind | outputScript
    ${'P2PKH'}  | ${false}    | ${Scripts.p2pkhOutput(crypto.hash160(publicKey))}
    ${'P2PKH'}  | ${true}     | ${Scripts.p2pkhOutput(crypto.hash160(publicKey))}
    ${'P2SH'}   | ${false}    | ${Scripts.p2shOutput(publicKey)}
    ${'P2SH'}   | ${true}     | ${Scripts.p2shOutput(publicKey)}
    ${'P2WPKH'} | ${false}    | ${Scripts.p2wpkhOutput(crypto.hash160(publicKey))}
    ${'P2WPKH'} | ${true}     | ${Scripts.p2wpkhOutput(crypto.hash160(publicKey))}
    ${'P2WSH'}  | ${false}    | ${Scripts.p2wshOutput(publicKey)}
    ${'P2WSH'}  | ${true}     | ${Scripts.p2wshOutput(publicKey)}
    ${'P2TR'}   | ${false}    | ${Scripts.p2trOutput(toXOnly(publicKey))}
    ${'P2TR'}   | ${true}     | ${Scripts.p2trOutput(toXOnly(publicKey))}
  `(
    'should encode $name address (confidential: $shouldBlind)',
    ({ outputScript, shouldBlind }) => {
      expect(wallet.encodeAddress(outputScript, shouldBlind)).toMatchSnapshot();
    },
  );

  test('should return empty string as address for an empty script', () => {
    expect(wallet.encodeAddress(Buffer.alloc(0), true)).toEqual('');
    expect(wallet.encodeAddress(Buffer.alloc(0), false)).toEqual('');
  });

  test('should return empty string as address for scripts that cannot be encoded', () => {
    const outputScript = Buffer.from([ops.OP_RETURN, 1, 2, 3]);
    expect(wallet.encodeAddress(outputScript, true)).toEqual('');
    expect(wallet.encodeAddress(outputScript, false)).toEqual('');
  });

  test('should blind by default', () => {
    expect(
      wallet
        .encodeAddress(Scripts.p2trOutput(toXOnly(publicKey)))
        .startsWith(Networks.liquidRegtest.blech32),
    ).toEqual(true);
  });
});
