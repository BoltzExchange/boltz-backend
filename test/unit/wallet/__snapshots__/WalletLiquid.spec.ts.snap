// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletLiquid should derive blinding keys from script for address AzpvTi6t8GTVxhg6tZ4vxNCdQKS9YYMMgJmxQAFYfhVvaogb6iVgTixvtv246tSbeM3zdgG1Z2ToreMt 1`] = `
{
  "data": [
    106,
    224,
    188,
    246,
    172,
    239,
    156,
    229,
    189,
    135,
    73,
    11,
    203,
    152,
    88,
    222,
    36,
    241,
    201,
    64,
    47,
    31,
    102,
    79,
    140,
    247,
    159,
    225,
    105,
    38,
    82,
    74,
  ],
  "type": "Buffer",
}
`;

exports[`WalletLiquid should derive blinding keys from script for address CTEvqk9mbKSWnkPhF7DwqHf5X5Jx1Q25LXh4sprdqj4KRgMZTZaiGrhCCDfWrDyVqBbxUrhyCtLwgB7J 1`] = `
{
  "data": [
    9,
    92,
    208,
    181,
    55,
    69,
    139,
    221,
    246,
    80,
    157,
    160,
    170,
    232,
    140,
    154,
    1,
    116,
    134,
    90,
    46,
    200,
    30,
    67,
    0,
    44,
    174,
    107,
    164,
    253,
    26,
    160,
  ],
  "type": "Buffer",
}
`;

exports[`WalletLiquid should derive blinding keys from script for address el1qqwxdjljvzdukcfxeammq6jktrvcurh329k7j208pm4rwe7anu2luhh7255kde06j2et2d0g5ym6yy949rx6rkp04xeav62vjp 1`] = `
{
  "data": [
    171,
    75,
    129,
    24,
    207,
    84,
    235,
    215,
    147,
    250,
    37,
    130,
    147,
    233,
    213,
    160,
    19,
    41,
    33,
    225,
    54,
    79,
    71,
    225,
    253,
    123,
    142,
    144,
    95,
    102,
    250,
    29,
  ],
  "type": "Buffer",
}
`;

exports[`WalletLiquid should encode P2PKH address (confidential: false) 1`] = `"2df2n73X6jTPddy31rjMfgZhxC6bkcEU9XT"`;

exports[`WalletLiquid should encode P2PKH address (confidential: true) 1`] = `"CTEnqWfNvTGTtoYYQL6WiAvNcHDWBWa1rSYBfph2zsUGHBafMs2P8wetoiQSwzvyzQjZK12EzBgZpkLW"`;

exports[`WalletLiquid should encode P2SH address (confidential: false) 1`] = `"XGxGZfq18bFSncj9wc4g6WA5VPNprJdAmg"`;

exports[`WalletLiquid should encode P2SH address (confidential: true) 1`] = `"Azppnmg47wJVTyJvBWUR4vkgZQ6gkK2VkJJrwsUbiLWGmFMWZDScTd6NmrX3LND5yM2UE9SGesLs5KfX"`;

exports[`WalletLiquid should encode P2TR address (confidential: false) 1`] = `"ert1p7qypc2gpr437ws0yhl3yvk57rweq8pfdyw04g8vjmjxeus9ak0nqwf79lm"`;

exports[`WalletLiquid should encode P2TR address (confidential: true) 1`] = `"el1pq2w6dcmt9smmce3ear69m9yuk328gm6hx05vyk93uutv0ugzx7mp0uqgrs5sz8truaq7f0lzgedfuxajqwzj6gul2swe9hydneqtmvlxvadnu96fepd2"`;

exports[`WalletLiquid should encode P2WPKH address (confidential: false) 1`] = `"ert1q84ufadu3juwu6zjcdcweqhe9ewhwxrvtj3lmzj"`;

exports[`WalletLiquid should encode P2WPKH address (confidential: true) 1`] = `"el1qqdx8446sv2hzmnphuzkt2ct848n05zt6r6ue72nl56l6xjf2p0exk0tcn6mer9cae599smsajp0jtjawuvxcklqsf6r8tdrqz"`;

exports[`WalletLiquid should encode P2WSH address (confidential: false) 1`] = `"ert1qxxywq89mszpuv6t4puyfypf8qafcdg4zcwed69ku4sr6tlg0tteq7kr8c6"`;

exports[`WalletLiquid should encode P2WSH address (confidential: true) 1`] = `"el1qqfq4vetpg6ly3dapjcltgyg0475nqcq4592tcvuc25t2t0y84s692vvguqwthqyrce5h2rcgjgzjwp6ns6329sajm5tdetq85h7s7khj388xz2mq4qgg"`;
